### DISCUSSION TITLE
Is there some guidance on when to use `setup()` vs `userEvent.whatever`?

### DISCUSSION THREAD

**Discussion by camjackson** [testing-library/user-event#1036] [Q&A] [answered]
I've been using testing-library and user-event for a few years. I've always done `userEvent.click()`, and never had any problems with it.

I recently started on a new team and have found that they do

```
const user = userEvent.setup();
user.click();
```

I had never seen this before, went and did a bunch of reading, [including this page of the docs](https://testing-library.com/docs/user-event/setup/), and some of the issues + PRs discussing the implementation. It seems that `setup()` is important if you want proper simulation of multi-step inputs like shift+click, but that is surely a rare case? >99% of my tests never have to worry about that, and work just fine with `userEvent.click()`, as far as I can tell.

I find `setup()` to be a bit boilerplatey. I don't want to have to add `const user = userEvent.setup()` to the start of every single test (or worse, use `let` and `beforeEach`). It just feels messy. But I will do it if there is a good reason. The docs have this to say:

> You can also call the APIs directly on the default export. This will call setup internally and then use the method on the instance.

> This exists to ease the transition to version 14 and writing simple tests. Using the methods on the instances returned by userEvent.setup() is recommended.

What I really want to know is: if I'm not doing tricky multi-step interactions or trying to configure userEvent to do anything special, is there any reason to use `setup()`? Or can I continue on without the extra step as I always have? Will there be a performance penalty from `setup` being invoked on every interaction rather than once at the start of my test? Or is the difference negligible?

**Comment by ph-fritsche** [comment, marked as answer]
> This exists to ease the transition to version 14 and writing simple tests. Using the methods on the instances returned by `userEvent.setup()` is recommended.

This pretty much sums it up.
You *should* use `userEvent.setup()` after creating the document and before you `render()` your component.
You *can* keep on using the direct APIs if rewriting tests is to much work on a project or if you only write a simple (throw-away) test.

In almost all cases it will not make a difference, but if you hit one of the rare edge cases, it might be hard to debug - especially since most environments don't create a new document for each test. So the first test will apply `.setup()` after rendering and every subsequent test will have our workarounds applied from the test before and only create a new input device state after rendering.

Yes, boilerplate in every test can be messy. A `setup` function that you can share across many tests is the way to go.
https://testing-library.com/docs/user-event/intro#writing-tests-with-userevent

You can write a test helper that allows you to create single-line arrange steps in your tests.
See for example the test helpers on this repo: https://github.com/testing-library/user-event/blob/main/tests/_helpers/setup.ts

**Reply by camjackson** [reply]
Thanks, that's helpful. I'm still a bit unconvinced on the value of `setup` if it will almost never make a difference. But knowing that I can avoid the boilerplate by wrapping `setup` and `render` into a single step makes me a lot happier. Cheers.

**Reply by neriyarden** [reply]
It would be nice to have an eslint rule to enforce using `userEvent.setup()` in the [eslint-plugin-testing-library](https://testing-library.com/docs/ecosystem-eslint-plugin-testing-library/)

**Reply by minh-nguyen-97** [reply]
I agree with @neriyarden that it would be great to have ESLint rule for this since it's quite boilerplate.
Looks like there is [a request raised for that](https://github.com/testing-library/eslint-plugin-testing-library/issues/646)

**Reply by Stan-Stani** [reply]
Another problem is https://github.com/testing-library/eslint-plugin-testing-library/issues/812 which gives me another reason not to adopt the recommended pattern of using `userEvent.setup()` as I lose out on ESlint ensuring I'm properly awaiting methods.

*ETA*: I was able to get a partial solution by adding the rule `"@typescript-eslint/no-floating-promises": "warn"`, and following the guide to enable typed linting: https://typescript-eslint.io/getting-started/typed-linting/

**Reply by dougbacelar** [reply]
Completely agree with @camjackson here.

After that update repos got flooded with `userEvent.setup` everywhere, and I am yet to encounter a case where it is needed. I have stopped using it and it works fine. If I ever face an issue, I will use the test helper mentioned above, which looks like a good compromise.

This newer answer also really helped me understand the difference in behaviour better. https://github.com/testing-library/user-event/discussions/1207#discussioncomment-9649749

**Reply by camjackson** [reply]
FWIW, in the three years since asking the question, I have used `setup` exactly 0 times, and I have had no issues.

I'm just going to keep using `userEvent.click()` because it works fine and it's so much cleaner.

**Reply by exhilaratedguy** [reply]
I have followed the recommendation by having a custom `customRender()` function that basically returns `{ user: userEvent.setup(), ...render() }` in a new project and it seems to working fine for now. I have not bothered to update older projects though.

I was already using a `customRender()` function to wrap the JSX with Lingui's provider though, so it didn't feel as if I was creating something that would require me to change every render import on every test for such a small thing.
